sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala

```diff
@@ -0,0 +1,245 @@
+ */
+
+package org.apache.spark.sql.execution.window
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Ascending, Attribute, DenseRank, Expression, Rank, RowNumber, SortOrder, UnsafeProjection, UnsafeRow}
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateOrdering
+import org.apache.spark.sql.catalyst.plans.physical.{AllTuples, ClusteredDistribution, Distribution, Partitioning}
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+
+sealed trait WindowGroupLimitMode
+
+case object Partial extends WindowGroupLimitMode
+
+case object Final extends WindowGroupLimitMode
+
+/**
+ * This operator is designed to filter out unnecessary rows before WindowExec
+ * for top-k computation.
+ * @param partitionSpec Should be the same as [[WindowExec#partitionSpec]].
+ * @param orderSpec Should be the same as [[WindowExec#orderSpec]].
+ * @param rankLikeFunction The function to compute row rank, should be RowNumber/Rank/DenseRank.
+ * @param limit The limit for rank value.
+ * @param mode The mode describes [[WindowGroupLimitExec]] before or after shuffle.
+ * @param child The child spark plan.
+ */
+case class WindowGroupLimitExec(
+    partitionSpec: Seq[Expression],
+    orderSpec: Seq[SortOrder],
+    rankLikeFunction: Expression,
+    limit: Int,
+    mode: WindowGroupLimitMode,
+    child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def requiredChildDistribution: Seq[Distribution] = mode match {
+    case Partial => super.requiredChildDistribution
+    case Final =>
+      if (partitionSpec.isEmpty) {
+        AllTuples :: Nil
+      } else {
+        ClusteredDistribution(partitionSpec) :: Nil
+      }
+  }
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec)
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  protected override def doExecute(): RDD[InternalRow] = rankLikeFunction match {
+    case _: RowNumber =>
+      child.execute().mapPartitionsInternal(
+        SimpleGroupLimitIterator(partitionSpec, output, _, limit))
+    case _: Rank =>
+      child.execute().mapPartitionsInternal(
+        RankGroupLimitIterator(partitionSpec, output, _, orderSpec, limit))
+    case _: DenseRank =>
+      child.execute().mapPartitionsInternal(
+        DenseRankGroupLimitIterator(partitionSpec, output, _, orderSpec, limit))
+  }
+
+  override protected def withNewChildInternal(newChild: SparkPlan): WindowGroupLimitExec =
+    copy(child = newChild)
+}
+
+abstract class WindowIterator extends Iterator[InternalRow] {
+
+  def partitionSpec: Seq[Expression]
+
+  def output: Seq[Attribute]
+
+  def input: Iterator[InternalRow]
+
+  def limit: Int
+
+  val grouping = UnsafeProjection.create(partitionSpec, output)
```

Comment:
I'm wondering if we should optimize for the special case where the function is row_number and the partition spec is empty. For this case, we just need to do a top n, no need to sort the entire input.

Reply:
We can do this later if there is no TPC query hitting it.

Reply:
Yes. I have another PR do the work.